kvbc/src/categorization/kv_blockchain.cpp

```diff
@@ -411,38 +416,42 @@ BlockMerkleOutput KeyValueBlockchain::handleCategoryUpdates(BlockId block_id,
                                                             categorization::RawBlock& raw_block) {
   // if true means that category is new and we should create an instance of it.
   if (insertCategoryMapping(category_id, detail::CATEGORY_TYPE::merkle, write_batch)) {
-    if (const auto [itr, inserted] = categorires_.try_emplace(category_id, BlockMerkleCategory{}); !inserted) {
-      (void)itr;
+    if (const auto [_, inserted] = categorires_.try_emplace(category_id, detail::BlockMerkleCategory{native_client_});
+        !inserted) {
+      (void)_;
       throw std::runtime_error{"Category already exists = " + category_id};
     }
   }
 
-  BlockMerkleOutput mui;
-  for (auto& [k, v] : updates.kv) {
-    (void)v;
-    mui.keys[k] = MerkleKeyFlag{false};
-  }
-  for (auto& k : updates.deletes) {
-    mui.keys[k] = MerkleKeyFlag{true};
+  auto itr = categorires_.find(category_id);
+  if (itr == categorires_.end()) {
+    throw std::runtime_error{"Category is not present in memory = " + category_id};
   }
-  raw_block.data.category_root_hash[category_id] = mui.root_hash;
-  return mui;
+  LOG_DEBUG(CAT_BLOCK_LOG, "Adding updates of block [" << block_id << "] to the BlockMerkleCategory");
+  return std::get<detail::BlockMerkleCategory>(itr->second).add(block_id, std::move(updates), write_batch);
 }
 
 VersionedOutput KeyValueBlockchain::handleCategoryUpdates(BlockId block_id,
                                                           const std::string& category_id,
                                                           VersionedInput&& updates,
                                                           concord::storage::rocksdb::NativeWriteBatch& write_batch,
                                                           categorization::RawBlock& raw_block) {
-  VersionedOutput out;
-  for (auto& [k, v] : updates.kv) {
-    (void)v;
-    out.keys[k] = VersionedKeyFlags{false, v.stale_on_update};
+  // if true means that category is new and we should create an instance of it.
+  if (insertCategoryMapping(category_id, detail::CATEGORY_TYPE::versioned_kv, write_batch)) {
+    if (const auto [_, inserted] =
+            categorires_.try_emplace(category_id, detail::VersionedKeyValueCategory{category_id, native_client_});
+        !inserted) {
+      (void)_;
+      throw std::runtime_error{"Category already exists = " + category_id};
```

Comment:
I think we should assert in this case as this should never happen. Please correct me if I am wrong.

Reply:
I looked at assertions and exceptions as having the same behavior eventually.
I'll change it.